source/kernel/src/bbq.rs

```diff
@@ -84,14 +103,166 @@ impl Drop for BBQStorage {
     }
 }
 
-impl BBQBidiHandle {
+use abi::bbqueue_ipc::{
+    GrantR as InnerGrantR,
+    GrantW as InnerGrantW,
+};
+
+pub struct GrantW {
+    grant: InnerGrantW<'static>,
+    // When we commit, we trigger OUR commit waitcell
+    our_commit: NonNull<WaitCell>,
+}
+
+impl Deref for GrantW {
+    type Target = [u8];
+
+    #[inline(always)]
+    fn deref(&self) -> &Self::Target {
+        self.grant.deref()
+    }
+}
+
+impl DerefMut for GrantW {
+    #[inline(always)]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.grant.deref_mut()
+    }
+}
+
+impl GrantW {
+    pub fn commit(self, used: usize) {
+        self.grant.commit(used);
+        // If we freed up any space, notify the waker on the reader side
+        if used != 0 {
+            unsafe {
+                self.our_commit.as_ref().notify();
+            }
+        }
+    }
+}
+
+pub struct GrantR {
+    grant: InnerGrantR<'static>,
+    our_release: NonNull<WaitCell>,
+}
+
+impl Deref for GrantR {
+    type Target = [u8];
+
     #[inline(always)]
-    pub fn producer(&self) -> &Producer<'static> {
-        &self.producer
+    fn deref(&self) -> &Self::Target {
+        self.grant.deref()
     }
+}
 
+impl DerefMut for GrantR {
     #[inline(always)]
-    pub fn consumer(&self) -> &Consumer<'static> {
-        &self.consumer
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.grant.deref_mut()
+    }
+}
+
+impl GrantR {
+    pub fn release(self, used: usize) {
+        self.grant.release(used);
+        // If we freed up any space, notify the waker on the reader side
+        if used != 0 {
+            unsafe {
+                self.our_release.as_ref().notify();
+            }
+        }
+    }
+}
+
+impl BBQBidiHandle {
+    // async fn send_grant(buf_len: usize) -> GrantW
+    // async fn read_grant() -> GrantR
+    pub async fn send_grant_max(&self, max: usize) -> GrantW {
+        loop {
+            match self.producer.grant_max_remaining(max) {
+                Ok(wgr) => {
+                    trace!(
+                        size = wgr.len(),
+                        max = max,
+                        "Got bbqueue max write grant",
+                    );
+                    let waker = unsafe {
+                        let ptr = addr_of_mut!((*self.our_waitcells.as_ptr()).commit_waitcell);
+                        NonNull::new_unchecked(ptr)
+                    };
```

Comment:
whew, you love 2 see it

Reply:
Fixed this in [31d9370](https://github.com/tosc-rs/mnemos/pull/3/commits/31d93705ffd13380b0f1be85ef7f3061ba594fad)